Ram usage problem while encoding 4k videos  (OpenCL Scaling Memory Leak)

### Please describe the problem or feature request in detail:
I'm using Handbrake to encode a large quantity of 4k videos from a Youtube Project. There are 50 files in queue. 
I'm not sure if I've well understood the issue: after every encoding ends (the videos are really short, less than 30 sec each) the Ram looks to be not cleaned. I've got only 8gb of Ram and, for this, after 3-4 encoding Windows crash while Handbrake is still trying to allocate more ram. During the first encoding the Ram usage is less than 5000mb, during the second less than 6000, during the third less than 7000, during the fourth (if it doesn't crash) reach 7800mb.

Edit: To avoid the crash I have to close Handbrake before the forth encode (if I just stop the queue the Ram usage stay the same) and reopen it.

### What are the steps to reproduce this problem: 
Make a batch encoding of more than four 4k videos (x265@8000kbps + aac@112kbps)

### What version of HandBrake you are running:
Handbrake latest release - 1.0.1 (2016122900) 64bit

### What operating system and version and you running (e.g. OSX 10.11,  Windows 7, Ubuntu 14):
Windows 10, no more programs running in background than antivirus. 
Hardware: fx8320e@3.6ghz, 8gb@1866mhz

### Was there any error message or error dialog, if so please detail or provide screenshots: 
I do not get any error message. If Windows does not crash it close every program. 

### Please provide the full activity log for the encode or scan attempt. You may attach the log as a file, post a pastebin URL to the log, or place the log inline below:
``` HandBrake 1.0.1 (2016122900) - 64bit
OS: Microsoft Windows NT 10.0.14393.0 - 64bit
CPU: AMD FX-8320E Eight-Core Processor              
Ram: 8091 MB, 
GPU Information:
  AMD Radeon R9 200 Series - 21.19.407.0
Screen: 2560x1080
Temp Dir: C:\Users\Marco\AppData\Local\Temp\
Install Dir: C:\Program Files\Handbrake
Data Dir: C:\Users\Marco\AppData\Roaming\HandBrake Team\HandBrake\1.0.1.0

-------------------------------------------


# Starting Encode ...

[16:42:18] hb_init: starting libhb thread
[16:42:18] 1 job(s) to process
[16:42:18] json job:
{
  "Audio": {
    "AudioList": [
      {
        "Bitrate": 112,
        "DRC": 0.0,
        "Encoder": 65536,
        "Gain": 0.0,
        "Mixdown": 4,
        "NormalizeMixLevel": false,
        "Samplerate": 0,
        "Name": "",
        "Track": 0,
        "DitherMethod": 0
      }
    ],
    "CopyMask": [
      1073807360,
      1073743872,
      1074003968,
      1073750016,
      1090519040,
      1074790400,
      1074266112,
      1107296256
    ],
    "FallbackEncoder": 2048
  },
  "Destination": {
    "ChapterList": [
      {
        "Name": "Chapter 1"
      }
    ],
    "ChapterMarkers": true,
    "File": "D:\\Risorse\\Video\\Progetti\\YouTube\\Video\\Redmi 4 Pro Storage\\Source\\20170104_191112.mp4",
    "Mp4Options": {
      "IpodAtom": false,
      "Mp4Optimize": false
    },
    "Mux": 131072
  },
  "Filters": {
    "FilterList": [
      {
        "ID": 4,
        "Settings": {
          "mode": "7"
        }
      },
      {
        "ID": 3,
        "Settings": {
          "block-height": "16",
          "block-thresh": "40",
          "block-width": "16",
          "filter-mode": "2",
          "mode": "3",
          "motion-thresh": "1",
          "spatial-metric": "2",
          "spatial-thresh": "1"
        }
      },
      {
        "ID": 11,
        "Settings": {
          "crop-bottom": "0",
          "crop-left": "0",
          "crop-right": "0",
          "crop-top": "0",
          "height": "2160",
          "width": "3840"
        }
      },
      {
        "ID": 6,
        "Settings": {
          "mode": "2",
          "rate": "27000000/900900"
        }
      }
    ]
  },
  "PAR": {
    "Num": 1,
    "Den": 1
  },
  "Metadata": {},
  "SequenceID": 0,
  "Source": {
    "Angle": 1,
    "Range": {
      "Type": "chapter",
      "Start": 1,
      "End": 1
    },
    "Title": 20,
    "Path": "D:\\Risorse\\Video\\Progetti\\YouTube\\Video\\Redmi 4 Pro\\Source\\20170104_191112.mp4"
  },
  "Subtitle": {
    "Search": {
      "Burn": true,
      "Default": false,
      "Enable": true,
      "Forced": false
    },
    "SubtitleList": []
  },
  "Video": {
    "Encoder": 4096,
    "Level": "auto",
    "Bitrate": 8000,
    "TwoPass": false,
    "Turbo": false,
    "ColorMatrixCode": 0,
    "Options": "",
    "Preset": "faster",
    "Profile": "auto",
    "OpenCL": true,
    "HWDecode": false,
    "QSV": {
      "Decode": false,
      "AsyncDepth": 0
    }
  }
}
[16:42:18] CPU:
[16:42:18]  - logical processor count: 8
[16:42:18] OpenCL device #1: Advanced Micro Devices, Inc. Tahiti
[16:42:18]  - OpenCL version: 1.2 AMD-APP (2236.10)
[16:42:18]  - driver version: 2236.10
[16:42:18]  - device type:    GPU
[16:42:18]  - supported:      YES
[16:42:18] Intel Quick Sync Video support: no
[16:42:18] hb_scan: path=D:\Risorse\Video\Progetti\YouTube\Video\Redmi 4 Pro\Source\20170104_191112.mp4, title_index=20
udfread ERROR: ECMA 167 Volume Recognition failed
src/libbluray/disc/disc.c:274: failed opening UDF image D:\Risorse\Video\Progetti\YouTube\Video\Redmi 4 Pro\Source\20170104_191112.mp4
src/libbluray/disc/disc.c:352: error opening file BDMV\index.bdmv
src/libbluray/disc/disc.c:352: error opening file BDMV\BACKUP\index.bdmv
[16:42:18] bd: not a bd - trying as a stream/file instead
libdvdread: Encrypted DVD support unavailable.
libdvdread:DVDOpenFileUDF:UDFFindFile /VIDEO_TS/VIDEO_TS.IFO failed
libdvdread:DVDOpenFileUDF:UDFFindFile /VIDEO_TS/VIDEO_TS.BUP failed
libdvdread: Can't open file VIDEO_TS.BUP.
[16:42:18] dvd: not a dvd - trying as a stream/file instead
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from 'D:\Risorse\Video\Progetti\YouTube\Video\Redmi 4 Pro\Source\20170104_191112.mp4':
  Metadata:
    major_brand     : mp42
    minor_version   : 0
    compatible_brands: isommp42
    creation_time   : 2017-01-04 18:12:04
  Duration: 00:00:51.55, start: 0.000000, bitrate: 48300 kb/s
    Stream #0:0(eng): Video: h264 (High) [avc1 / 0x31637661]
      yuv420p, 3840x2160, 48040 kb/s
      30 fps, 90k tbn (default)
    Metadata:
      creation_time   : 2017-01-04 18:12:04
      handler_name    : VideoHandle
    Stream #0:1(eng): Audio: aac (LC) [mp4a / 0x6134706D]
      48000 Hz, stereo, fltp, 256 kb/s (default)
    Metadata:
      creation_time   : 2017-01-04 18:12:04
      handler_name    : SoundHandle
[16:42:19] scan: decoding previews for title 20
[16:42:20] scan: audio 0x1: aac, rate=48000Hz, bitrate=256007 English (AAC) (2.0 ch)
[16:42:27] scan: 10 previews, 3840x2160, 30.007 fps, autocrop = 0/0/0/0, aspect 16:9, PAR 1:1
[16:42:27] scan: supported video decoders: avcodec qsv
[16:42:27] libhb: scan thread found 1 valid title(s)
[16:42:27] Skipping subtitle scan.  No suitable subtitle tracks.
[16:42:27] starting job
[16:42:27] decomb filter thread started for segment 0
[16:42:27] decomb filter thread started for segment 4
[16:42:27] decomb filter thread started for segment 2
[16:42:27] decomb filter thread started for segment 3
[16:42:27] decomb filter thread started for segment 1
[16:42:27] decomb filter thread started for segment 5
[16:42:27] decomb filter thread started for segment 6
[16:42:27] decomb filter thread started for segment 7
[16:42:27] decomb check thread started for segment 0
[16:42:27] decomb check thread started for segment 1
[16:42:27] decomb check thread started for segment 2
[16:42:27] decomb check thread started for segment 3
[16:42:27] decomb check thread started for segment 4
[16:42:27] decomb check thread started for segment 5
[16:42:27] mask erode thread started for segment 3
[16:42:27] decomb check thread started for segment 7
[16:42:27] mask filter thread started for segment 0
[16:42:27] mask filter thread started for segment 1
[16:42:27] mask filter thread started for segment 2
[16:42:27] mask filter thread started for segment 3
[16:42:27] mask filter thread started for segment 4
[16:42:27] mask filter thread started for segment 5
[16:42:27] mask filter thread started for segment 6
[16:42:27] mask filter thread started for segment 7
[16:42:27] yadif thread started for segment 3
[16:42:27] mask erode thread started for segment 1
[16:42:27] mask erode thread started for segment 2
[16:42:27] decomb check thread started for segment 6
[16:42:27] mask erode thread started for segment 4
[16:42:27] work: track 1, dithering not supported by codec
[16:42:27] mask erode thread started for segment 5
[16:42:27] mask erode thread started for segment 6
[16:42:27] work: only 1 chapter, disabling chapter markers
[16:42:27] job configuration:
[16:42:27] mask erode thread started for segment 7
[16:42:27]  * source
[16:42:27]    + D:\Risorse\Video\Progetti\YouTube\Video\Redmi 4 Pro\Source\20170104_191112.mp4
[16:42:27]    + title 20, chapter(s) 1 to 1
[16:42:27]    + container: mov,mp4,m4a,3gp,3g2,mj2
[16:42:27]    + data rate: 48300 kbps
[16:42:27]  * destination
[16:42:27]    + D:\Risorse\Video\Progetti\YouTube\Video\Redmi 4 Pro Storage\Source\20170104_191112.mp4
[16:42:27]    + container: MPEG-4 (libavformat)
[16:42:27] mask dilate thread started for segment 0
[16:42:27]  * video track
[16:42:27]    + decoder: h264
[16:42:27]      + bitrate 48040 kbps
[16:42:27]    + filters
[16:42:27]      + Comb Detect (mode=3:spatial-metric=2:motion-thresh=1:spatial-thresh=1:filter-mode=2:block-thresh=40:block-width=16:block-height=16)
[16:42:27]      + Decomb (mode=39)
[16:42:27]      + Framerate Shaper (mode=2:rate=27000000/900900)
[16:42:27] mask dilate thread started for segment 1
[16:42:27] mask dilate thread started for segment 2
[16:42:27]        + frame rate: 30.007 fps -> peak rate limited to 29.970 fps
[16:42:27]      + Crop and Scale (width=3840:height=2160:crop-top=0:crop-bottom=0:crop-left=0:crop-right=0)
[16:42:27] mask dilate thread started for segment 3
[16:42:27]        + source: 3840 * 2160, crop (0/0/0/0): 3840 * 2160, scale: 3840 * 2160
[16:42:27]    + Output geometry
[16:42:27]      + storage dimensions: 3840 x 2160
[16:42:27]      + pixel aspect ratio: 1 : 1
[16:42:27]      + display dimensions: 3840 x 2160
[16:42:27]    + encoder: H.265 (libx265)
[16:42:27]      + preset:  faster
[16:42:27] mask dilate thread started for segment 4
[16:42:27] mask dilate thread started for segment 5
[16:42:27]      + profile: auto
[16:42:27]      + bitrate: 8000 kbps, pass: 0
[16:42:27]  * audio track 1
[16:42:27]    + decoder: English (AAC) (2.0 ch) (track 1, id 0x1)
[16:42:27]      + bitrate: 256 kbps, samplerate: 48000 Hz
[16:42:27]    + mixdown: Stereo
[16:42:27] mask dilate thread started for segment 6
[16:42:27]    + encoder: AAC (libavcodec)
[16:42:27]      + bitrate: 112 kbps, samplerate: 48000 Hz
[16:42:27] mask dilate thread started for segment 7
[16:42:27] yadif thread started for segment 0
[16:42:27] yadif thread started for segment 1
[16:42:27] yadif thread started for segment 2
[16:42:27] mask erode thread started for segment 0
[16:42:27] yadif thread started for segment 4
[16:42:27] yadif thread started for segment 5
[16:42:27] yadif thread started for segment 6
[16:42:27] yadif thread started for segment 7
[16:42:28] sync: expecting 1547 video frames
x265 [info]: HEVC encoder version 2.1
x265 [info]: build info [Windows][GCC 5.4.0][64 bit] 8bit
x265 [info]: using cpu capabilities: MMX2 SSE2Fast SSSE3 SSE4.2 AVX XOP FMA4 FMA3 LZCNT BMI1
x265 [info]: Main profile, Level-5 (Main tier)
x265 [info]: Thread pool created using 8 threads
x265 [info]: Slices                              : 1
x265 [info]: frame threads / pool features       : 3 / wpp(34 rows)
x265 [info]: Coding QT: max CU size, min CU size : 64 / 8
x265 [info]: Residual QT: max TU size, max depth : 32 / 1 inter / 1 intra
x265 [info]: ME / range / subpel / merge         : hex / 57 / 2 / 2
x265 [info]: Keyframe min / max / scenecut       : 30 / 300 / 40
x265 [info]: Lookahead / bframes / badapt        : 15 / 4 / 0
x265 [info]: b-pyramid / weightp / weightb       : 1 / 1 / 0
x265 [info]: References / ref-limit  cu / depth  : 2 / on / on
x265 [info]: AQ: mode / str / qg-size / cu-tree  : 1 / 1.0 / 32 / 1
x265 [info]: Rate Control / qCompress            : ABR-8000 kbps / 0.60
x265 [info]: tools: rd=2 psy-rd=2.00 early-skip rskip signhide tmvp fast-intra
x265 [info]: tools: strong-intra-smoothing lslices=8 deblock sao
[16:42:28] sync: first pts video is 0
[16:42:28] sync: "Chapter 1" (1) at frame 1 time 0
[16:42:28] sync: first pts audio 0x1 is 0
x265 [error]: malloc of size 2466816 failed
x265 [error]: memory allocation failure, aborting encode
  ```

--- a/libhb/cropscale.c
+++ b/libhb/cropscale.c
@@ -142,6 +142,10 @@ static void hb_crop_scale_close( hb_filter_object_t * filter )
         {
             HB_OCL_BUF_FREE(hb_ocl, pv->os->bicubic_x_weights);
             HB_OCL_BUF_FREE(hb_ocl, pv->os->bicubic_y_weights);
+            if (pv->os->initialized == 1)
+            {
+                hb_ocl->clReleaseKernel(pv->os->m_kernel);
+            }
         }
         free(pv->os);
     }

--- a/libhb/openclwrapper.c
+++ b/libhb/openclwrapper.c
@@ -1126,6 +1126,11 @@ int hb_cl_free_mapped_buffer(cl_mem mem, unsigned char *addr)
         hb_ocl->clWaitForEvents(1, &event);
     else
         hb_log("hb_free_mapped_buffer: error %d", status);
+
+    status = hb_ocl->clReleaseMemObject(mem);
+    if (status != CL_SUCCESS)
+        hb_log("hb_free_mapped_buffer: release error %d",status);
+
     return (status == CL_SUCCESS) ? 1 : 0;
 }
 

--- a/libhb/work.c
+++ b/libhb/work.c
@@ -1831,6 +1831,12 @@ cleanup:
     }
 
     hb_buffer_pool_free();
+
+    if (job->use_opencl)
+    {
+        hb_release_opencl_run_env();
+    }
+
     hb_job_close(&job);
 }
 

